packages/edit-navigation/src/store/actions.js

```diff
@@ -49,169 +38,311 @@ export function setSelectedMenuId( menuId ) {
  * @param {Object} post A navigation post to process
  * @return {Function} An action creator
  */
-export const createMissingMenuItems = serializeProcessing( function* ( post ) {
+export const createMissingMenuItems = ( post ) => async ( {
+	dispatch,
+	registry,
+} ) => {
 	const menuId = post.meta.menuId;
+	// @TODO: extract locks to a separate package?
+	const lock = await registry
+		.dispatch( 'core' )
+		.__unstableAcquireStoreLock( STORE_NAME, [ 'savingMenu' ], {
+			exclusive: false,
+		} );
+	try {
+		const menuItemIdToBlockId = await dispatch(
+			getEntityRecordIdToBlockIdMapping( post.id )
+		);
+		const knownBlockIds = new Set( Object.values( menuItemIdToBlockId ) );
 
-	const mapping = yield getMenuItemToClientIdMapping( post.id );
-	const clientIdToMenuId = invert( mapping );
-
-	const stack = [ post.blocks[ 0 ] ];
-	while ( stack.length ) {
-		const block = stack.pop();
-		if ( ! ( block.clientId in clientIdToMenuId ) ) {
-			const menuItem = yield apiFetch( {
-				path: `/__experimental/menu-items`,
-				method: 'POST',
-				data: {
-					title: 'Placeholder',
-					url: 'Placeholder',
-					menu_order: 0,
-				},
-			} );
-
-			mapping[ menuItem.id ] = block.clientId;
-			const menuItems = yield resolveMenuItems( menuId );
-			yield dispatch(
-				'core',
-				'receiveEntityRecords',
-				'root',
-				'menuItem',
-				[ ...menuItems, menuItem ],
-				menuItemsQuery( menuId ),
-				false
-			);
+		const blocks = blocksTreeToFlatList( post.blocks[ 0 ].innerBlocks );
+		for ( const { block } of blocks ) {
+			if ( ! knownBlockIds.has( block.clientId ) ) {
+				const menuItem = await dispatch(
+					createPlaceholderMenuItem( block, menuId )
+				);
+				menuItemIdToBlockId[ menuItem.id ] = block.clientId;
+			}
 		}
-		stack.push( ...block.innerBlocks );
+
+		dispatch( {
+			type: 'SET_MENU_ITEM_TO_CLIENT_ID_MAPPING',
```

Comment:
We should probably switch to storing the menu item ID in a block attribute (`__internalMenuItemId`) as client IDs aren't stable. Same as https://github.com/WordPress/gutenberg/pull/28379. Later, though 🙂

Reply:
Yeah let's follow up with another PR.